### PR TITLE
Add jwt authentication

### PR DESCRIPTION
**Pull request by dougmartin**
Adds JWT as a Devise authentication source.

There are two types of JWT tokens the portal can generate - symmetrically signed portal tokens and RSA signed tokens for use by Firebase.

Also adds an API endpoint to generate signed links to the log-puller to each class offering.

### CODE REVIEW

File: app/controllers/api/v1/classes_controller.rb [hunk 1]
```diff
@@ -56,14 +56,22 @@ def render_info(clazz)
           :last_name => teacher.user.last_name
         }
       },
-      :students => clazz.students.includes(:user).map {|student|
+      :students => clazz.students.includes(:user, :learners).map {|student|
```
Comment (dougmartin) [inline app/controllers/api/v1/classes_controller.rb]: Remove :learners - this was there because I was writing out the student endpoints here too but I removed it

File: config/initializers/core_extensions/devise/strategies/json_web_token.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Your actual JWT Strategy
```
Comment (dougmartin) [inline config/initializers/core_extensions/devise/strategies/json_web_token.rb]: Remove this needless comment

File: config/initializers/devise.rb [hunk 3]
```diff
@@ -1,245 +1,247 @@
...
-  # config.router_name = :my_engine
-  #
-  # When using omniauth, Devise cannot automatically set Omniauth path,
-  # so you need to do it manually. For the users scope, it would be:
-  # config.omniauth_path_prefix = "/my_engine/users/auth"
-  config.warden do |manager|
-    manager.failure_app = CustomFailure
-  end
-end
+# Use this hook to configure devise mailer, warden hooks and so forth.
```
Comment (dougmartin) [inline config/initializers/devise.rb]: Fix CRLF

File: lib/signed_jwt.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    {data: decoded[0], header: decoded[1]}
+  end
+
+  def self.create_firebase_token(user, firebase_app_name, expires_in=3600, claims={})
+    app = FirebaseApp.find_by_name(firebase_app_name)
+    raise SignedJWT::Error.new("Unknown firebase app name: #{firebase_app_name}") if app.nil?
+
+    now = Time.now.to_i
+    payload = {
+      alg:	'RS256',
```
Comment (dougmartin) [inline lib/signed_jwt.rb]: Fix tab expansion here

File: spec/libs/json_web_token_spec.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
+require 'spec_helper'
```
Comment (dougmartin) [inline spec/libs/json_web_token_spec.rb]: This is the problem spec - SignedJWT::create_portal_token expects a ENV variable for the hmac secret to exist.  I can set the ENV variable in the spec but it also needs to be set when the stubbed request is made so devise can check it and I can't figure out where to set that for rspec other than the app envionment variables file for test.

File: app/controllers/api/v1/classes_controller.rb [hunk 6]
```diff
@@ -56,14 +56,22 @@ def render_info(clazz)
...
         {
           :id => url_for(student.user),
           :email => student.user.email,
           :first_name => student.user.first_name,
           :last_name => student.user.last_name
         }
-      }
+      },
```
Comment (dougmartin) [inline app/controllers/api/v1/classes_controller.rb]: Revert this change to the file and use the offerings endpoint

File: app/controllers/portal/log_report_links_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
+class Portal::LogReportLinksController < ApplicationController
```
Comment (dougmartin) [inline app/controllers/portal/log_report_links_controller.rb]: Move this to the api controller folder

File: config/initializers/core_extensions/devise/strategies/json_web_token.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+# "config/initializers/core_extensions/devise/strategies/json_web_token.rb"
+
+module Devise
```
Comment (dougmartin) [inline config/initializers/core_extensions/devise/strategies/json_web_token.rb]: Add reference in users model like `:bearer_token_authenticatable` and move this to the /lib folder.

File: lib/signed_jwt.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  def self.hmac_algorithm
+    'HS256'
+  end
+
+  def self.rsa_algorithm
+    'RS256'
+  end
+
+  def self.hmac_secret
+    secret = ENV['JWT_HMAC_SECRET']
```
Comment (scytacki) [inline lib/signed_jwt.rb]: Please add a line in `config/app_environment_variables.sample.rb` documenting this environment variable. 
How long should the secret be?

**Comment by scytacki** [comment]
Looks good. 👍
